lib/puppet/configurer.rb

```diff
@@ -258,38 +270,40 @@ def run_internal(options)
     end
 
     begin
-      unless Puppet[:node_name_fact].empty?
-        query_options = get_facts(options)
-      end
+      query_options = get_facts(options) unless Puppet[:node_name_fact].empty?
 
       configured_environment = Puppet[:environment] if Puppet.settings.set_by_config?(:environment)
 
       # We only need to find out the environment to run in if we don't already have a catalog
       unless (options[:catalog] || Puppet[:strict_environment_mode])
         begin
-          if node = options[:node] || Puppet::Node.indirection.find(Puppet[:node_name_value],
+          node = nil
+          node_retr_time = thinmark do
+            node = options[:node] || Puppet::Node.indirection.find(Puppet[:node_name_value],
               :environment => Puppet::Node::Environment.remote(@environment),
               :configured_environment => configured_environment,
               :ignore_cache => true,
               :transaction_uuid => @transaction_uuid,
               :fail_on_404 => true)
+          end
+          options[:report].add_times(:node_retrieval, node_retr_time)
+          raise if node.nil?
```

Comment:
@joshcooper @jhelwig pointed out that we could still get nil if we get anything other than a 404 or 200 response https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/rest.rb#L190 so I added this line back

Reply:
Ok sounds good.